### PR TITLE
C++17 string_view test

### PR DESCRIPTION
**Pull request by janisozaur**
This is a dummy PR to test state of C++17 support.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -81,7 +81,7 @@ matrix:
           services:
             - docker
         - os: osx
-          osx_image: xcode8
+          osx_image: xcode9
```
Comment (janisozaur) [inline .travis.yml]: Xcode 9 is still in beta, latest from 8 line, 8.3, doesn't support C++17 at all.

File: src/openrct2-cli/CMakeLists.txt [hunk 2]
```diff
@@ -30,5 +30,5 @@ if (FORCE32)
     set(TARGET_M "-m32")
 endif ()
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu11 ${TARGET_M}")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++14 ${TARGET_M}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++17 ${TARGET_M}")
```
Comment (janisozaur) [inline src/openrct2-cli/CMakeLists.txt]: I think these override each other by being included from the top-level makefile, so perhaps some refactoring is needed here.

Comment (janisozaur) [inline src/openrct2-cli/CMakeLists.txt]: I will change it to `set(CMAKE_CXX_STANDARD 17)` later

**Comment by janisozaur** [comment]
Regarding `set(CMAKE_CXX_STANDARD 17)`: 17 is only available with cmake 3.8: https://cmake.org/cmake/help/v3.8/prop_tgt/CXX_STANDARD.html so I'd rather use what we have now until distros update their cmake versions.

**Comment by rwjuk** [comment]
> C++17 is not yet a thing

It's this, and the headaches that getting this working now across our supported platforms would cause, that leads me to the conclusion that we should wait until the standard is finalised and support is more widespread.

**Comment by IntelOrca** [comment]
Okay, given its not well supported yet we can look into this again next April.

**Comment by janisozaur** [comment]
> next April

Read: Ubuntu 18.04 LTS

**Comment by janisozaur** [comment]
I would like to have a release before we enable C++17 features.

Some platforms (e.g. Redox) don't fully support C++17 yet (the language may be supported, but not filesystem TS) and this would provide a fairly recent version to serve as a starting point for porting to those platforms.

**Comment by LRFLEW** [comment]
Since @janisozaur keeps referring to this when talking about switching to C++17, I'll go ahead and lay out the issue with it right now: Xcode support is kind of messed up.

Language-wise, it appears to support all the changes. (I can at least confirm that `if constexpr` works). Though, the included version of Clang doesn't support the `-std=c++17` flag, only the old `-std=c++1z` flag.

However, the main issue is the standard library that Xcode provides. As of Xcode 9.2, the standard library is still (mostly) stuck on C++14, and lacks a huge number of features. `<string_view>` is present, but it's about the only one. `any` and `optional` are available in the "experimental" variant, but pretty much everything else (including `<variant>` and `<filesystem>` are not present in any form.

This problem can be worked around by using an alternative, such as Boost (at least for macOS), but that will pose some challenges to the implementation, and will need to be taking into consideration before we make the switch to C++17.

**Comment by IntelOrca** [comment]
> This problem can be worked around by using an alternative, such as Boost (at least for macOS), but that will pose some challenges to the implementation, and will need to be taking into consideration before we make the switch to C++17.

Well in VS2017's case the filesystem namespace is still under `std::experimental` because they didn't want to upgrade it to `std` until the language spec was finalised which only happened in December. This means that if they don't do it by April then we will have to alias the namespace anyway which is presumably how you can use boost as a polyfill?

**Comment by janisozaur** [comment]
I hear GCC8 (due out in March/April, won't be a part of Ubuntu 18.04 _by default_) has filesystem moved out of `experimental`, I think it was first made available as early as GCC6.

Clang probably has it still in experimental, but all the targetted compilers should be tested.

**Comment by LRFLEW** [comment]
>Clang probably has it still in experimental, but all the targetted compilers should be tested.

Well, Xcode uses a fork of the LLVM project, so the feature lists are different. In the case of the Xcode version, there is NO implementation of filesystem, even in experimental.

And yes, we can use boost like a polyfill, but it's not simple. Importantly, you aren't allowed to modify the `::std` namespace, so we can't just make boost act like a standard library. The best chance we have to do that is to make our own namespace, and use `using` directives (so instead of `std::string_view`, we'd use `openrct2::string_view` and have `using std::string_view;`, `using std::experimental::string_view;`, or `using boost::string_view;` in the `::openrct2` namespace).

**Pull request closed without merging**